deployment/helm/crds/akri-configuration-crd.yaml

```diff
@@ -100,11 +100,20 @@ spec:
                               type: array
                               items:
                                 type: string
+                    coapcore: # {{CoAPCoREDiscoveryHandler}}
+                      type: object
+                      properties:
+                        ipAddresses: # {{DeviceIPs}}
+                          type: array
```

Comment:
Maybe add some filtering options so people can specify that they only want to discover devices that support certain [resource types](https://www.iana.org/assignments/core-parameters/core-parameters.xhtml) -- say "I only want to find printers".

Reply:
I've just added support to query filtering https://tools.ietf.org/html/rfc6690#section-4.1. This will allow to filter discovered resources by setting:

```yml
queryFilter: 
  name: rt
  value: oic.r.temperature
```

From my understanding of the standard, only 1 filter is allowed and the value itself is extremely limited. That's the rational for having a filter as fixed struct instead of list of filters. Devices are also not required to support filtering because of the constrained environment, so some filtering is done by the agent.

Reply:
Where in the Configuration would people pass their requested filters? Instead of using queryFilter and having to handle the case that devices might not support it, why not just filter the results in the DiscoveryHandler? So if the DiscoveryHandler discovers 3 devices [device1: temperature, brightness, and moisture sensors/resources; device2: moisture; device3: brightness], and the configuration requests moisture resources, then the DiscoveryHandler would filter out device3 and return the other 2.

Reply:
I'm currently passing `queryFilters` in the properties:

```yaml
properties:
  multicast: true
  multicastIpAddress: 224.0.1.187
  staticIpAddresses: []
  queryFilter:
    name: rt
    value: oic.r.temperature
  discoveryTimeoutSeconds: 5
```

> Instead of using queryFilter and having to handle the case that devices might not support it, why not just filter the results in the DiscoveryHandler

I'm already filtering the results nevertheless in the DiscoveryHandler, exactly because devices are not required to support filtering :) However, I preferred to support only the standard query filtering instead of some custom Akri filtering. This limits filtering because, as far as I understand the RFC, you can filter only one property and only one value. So you **cannot** write a filter like `rt=temperature,brightness`, only `rt=temperature` or `rt=brightness` or `rt=temp*`. Actually I haven't added support to wildcart yet because that would require an additional dependency to a regex crate, which I discovered is not in std in Rust. But I don't have anything against it.

The rational for supporting the standard filtering, although limited, is that it feels more suitable for a first iteration to keep things simple. When more uses will emerge, more custom configuration can be added. Otherwise, for the time being, it could feel counter-intuitive having custom filter rules because one might expect that they are transformed to a proper CoRE query but that cannot actually happen. Currently, I can instead happily see that `/well-known/core?rt=oic.r.temperature` is passed to my CoAP servers and the DiscoveryHandler filters the results anyway if the server ignores it. This is standard expected behaviour.

I can't deny I'm also biased because my first attempt was allowing a custom rule like `rt=temperature,brightness` but things can become quickly weird, both from user and Akri perspective. From the user developer perspective, one wonders how this is implemented since it's non-standard. The same doubt happens with any other kind of query structure. From Akri perspective, it become ugly to do query parsing and to do conversion between CoAP query and our custom query. For instance, if we allow a custom filter to allows only `temperature` and `brightness`, what would the CoAP request look like?

 - `/well-known/core?rt=oic.r.temperature` is wrong because it's missing the brightness type
 - `/well-known/core?rt=oic.r.brightness` has the reverse issue
 - `/well-known/core?rt=oic.r.temperature,oic.r.brightness` is not standard-compliant
 - `/well-known/core?rt=oic.r.temperature` and `/well-known/core?rt=oic.r.brightness`? That seems the only doable compromise, at the cost of duplicating each discovery request and increasing the overhead also on the edge device.

Therefore, in order to avoid complicating stuff without a clear use case for it, I think supporting the limited standard query filtering is better for the time being. There could be also the case that a future version of the CoAP standard allows slightly more complex queries and by sticking to the standard, we save a lot of headaches. What do you think?

Reply:
Starting with standard query filtering and then potentially doing custom filtering as a vnext sounds good. Maybe add that information to the proposal so we can keep track of it.